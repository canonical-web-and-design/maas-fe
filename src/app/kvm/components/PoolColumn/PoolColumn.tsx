import { useSelector } from "react-redux";

import { useZoneById } from "@/app/api/query/zones";
import DoubleRow from "@/app/base/components/DoubleRow";
import poolSelectors from "@/app/store/resourcepool/selectors";
import type {
  ResourcePool,
  ResourcePoolMeta,
} from "@/app/store/resourcepool/types";
import type { RootState } from "@/app/store/root/types";
import type { Zone, ZoneMeta } from "@/app/store/zone/types";

type Props = {
  poolId?: ResourcePool[ResourcePoolMeta.PK] | null;
  zoneId?: Zone[ZoneMeta.PK] | null;
};

const PoolColumn = ({ poolId, zoneId }: Props): JSX.Element | null => {
  const pool = useSelector((state: RootState) =>
    poolSelectors.getById(state, poolId)
  );
  const { data: zone } = useZoneById(zoneId);

  return (
    <DoubleRow
      primary={<span data-testid="zone">{zone?.name}</span>}
      secondary={<span data-testid="pool">{pool?.name}</span>}
    />
  );
};

export default PoolColumn;
