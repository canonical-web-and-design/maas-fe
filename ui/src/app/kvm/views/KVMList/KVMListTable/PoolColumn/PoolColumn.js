import React from "react";
import { useSelector } from "react-redux";

import {
  pod as podSelectors,
  resourcepool as poolSelectors,
  zone as zoneSelectors,
} from "app/base/selectors";
import DoubleRow from "app/base/components/DoubleRow";

const PoolColumn = ({ id }) => {
  const pod = useSelector((state) => podSelectors.getById(state, id));
  const pool = useSelector((state) =>
    poolSelectors.getById(state, pod && pod.pool)
  );
  const zone = useSelector((state) =>
    zoneSelectors.getById(state, pod && pod.zone)
  );

  return (
    <DoubleRow
      primary={<span data-test="pod-pool">{pool && pool.name}</span>}
      secondary={<span data-test="pod-zone">{zone && zone.name}</span>}
    />
  );
};

export default PoolColumn;
