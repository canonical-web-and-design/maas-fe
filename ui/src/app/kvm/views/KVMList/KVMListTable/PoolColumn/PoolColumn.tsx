import React from "react";
import { useSelector } from "react-redux";

import {
  pod as podSelectors,
  resourcepool as poolSelectors,
  zone as zoneSelectors,
} from "app/base/selectors";
import { RootState } from "app/store/root/types";
import DoubleRow from "app/base/components/DoubleRow";

type Props = { id: number };

const PoolColumn = ({ id }: Props): JSX.Element => {
  const pod = useSelector((state: RootState) =>
    podSelectors.getById(state, id)
  );
  const pool = useSelector((state: RootState) =>
    poolSelectors.getById(state, pod && pod.pool)
  );
  const zone = useSelector((state: RootState) =>
    zoneSelectors.getById(state, pod && pod.zone)
  );

  return (
    <DoubleRow
      primary={<span data-test="pod-pool">{pool && pool.name}</span>}
      secondary={<span data-test="pod-zone">{zone && zone.name}</span>}
    />
  );
};

export default PoolColumn;
