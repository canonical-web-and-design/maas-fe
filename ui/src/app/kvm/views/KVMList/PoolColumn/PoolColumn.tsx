import { useSelector } from "react-redux";

import DoubleRow from "app/base/components/DoubleRow";
import podSelectors from "app/store/pod/selectors";
import poolSelectors from "app/store/resourcepool/selectors";
import type { RootState } from "app/store/root/types";
import zoneSelectors from "app/store/zone/selectors";

type Props = { id: number };

const PoolColumn = ({ id }: Props): JSX.Element | null => {
  const pod = useSelector((state: RootState) =>
    podSelectors.getById(state, id)
  );
  const pool = useSelector((state: RootState) =>
    poolSelectors.getById(state, pod && pod.pool)
  );
  const zone = useSelector((state: RootState) =>
    zoneSelectors.getById(state, pod && pod.zone)
  );

  if (pod) {
    return (
      <DoubleRow
        primary={<span data-test="pod-pool">{pool?.name}</span>}
        secondary={<span data-test="pod-zone">{zone?.name}</span>}
      />
    );
  }
  return null;
};

export default PoolColumn;
