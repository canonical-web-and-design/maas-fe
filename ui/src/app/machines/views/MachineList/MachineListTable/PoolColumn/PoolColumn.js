import { Spinner } from "@canonical/react-components";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import { memo, useEffect, useState } from "react";

import { actions as machineActions } from "app/store/machine";
import machineSelectors from "app/store/machine/selectors";
import resourcePoolSelectors from "app/store/resourcepool/selectors";
import { useToggleMenu } from "app/machines/hooks";
import DoubleRow from "app/base/components/DoubleRow";
import poolsURLs from "app/pools/urls";

import { NodeActions } from "app/store/types/node";

export const PoolColumn = ({ onToggleMenu, systemId }) => {
  const dispatch = useDispatch();
  const [updating, setUpdating] = useState(null);
  const machine = useSelector((state) =>
    machineSelectors.getById(state, systemId)
  );
  const resourcePools = useSelector(resourcePoolSelectors.all);
  const toggleMenu = useToggleMenu(onToggleMenu, systemId);

  let poolLinks = resourcePools.filter((pool) => pool.id !== machine.pool.id);
  if (machine.actions.includes(NodeActions.SET_POOL)) {
    if (poolLinks.length !== 0) {
      poolLinks = poolLinks.map((pool) => ({
        children: pool.name,
        onClick: () => {
          dispatch(machineActions.setPool({ systemId, poolId: pool.id }));
          setUpdating(pool.id);
        },
      }));
    } else {
      poolLinks = [{ children: "No other pools available", disabled: true }];
    }
  } else {
    poolLinks = [
      { children: "Cannot change pool of this machine", disabled: true },
    ];
  }

  useEffect(() => {
    if (updating !== null && machine.pool.id === updating) {
      setUpdating(null);
    }
  }, [updating, machine.pool.id]);

  return (
    <DoubleRow
      menuLinks={onToggleMenu && poolLinks}
      menuTitle="Change pool:"
      onToggleMenu={toggleMenu}
      primary={
        <span data-test="pool">
          {updating !== null ? (
            <Spinner className="u-nudge-left--small" />
          ) : null}
          <Link className="p-link--soft" to={poolsURLs.pools}>
            {machine.pool.name}
          </Link>
        </span>
      }
      primaryAriaLabel="Pool"
      primaryTitle={machine.pool.name}
      secondary={
        <span title={machine.description} data-test="note">
          {machine.description}
        </span>
      }
      secondaryAriaLabel="Note"
      secondaryTitle={machine.description}
    />
  );
};

PoolColumn.propTypes = {
  onToggleMenu: PropTypes.func,
  systemId: PropTypes.string.isRequired,
};

export default memo(PoolColumn);
