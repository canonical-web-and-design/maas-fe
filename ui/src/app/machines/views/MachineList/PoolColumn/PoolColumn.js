import React from "react";
import PropTypes from "prop-types";
import { useSelector } from "react-redux";

import { machine as machineSelectors } from "app/base/selectors";
import DoubleRow from "app/base/components/DoubleRow";

const PoolColumn = ({ onToggleMenu, systemId }) => {
  const machine = useSelector(state =>
    machineSelectors.getBySystemId(state, systemId)
  );

  return (
    <DoubleRow
      onToggleMenu={onToggleMenu}
      primary={
        <span data-test="pool">
          <a className="p-link--soft" href="#/pools" title={machine.pool.name}>
            {machine.pool.name}
          </a>
        </span>
      }
      primaryAriaLabel="Pool"
      secondary={
        <span title={machine.description} data-test="note">
          {machine.description}
        </span>
      }
      secondaryAriaLabel="Note"
    />
  );
};

PoolColumn.propTypes = {
  onToggleMenu: PropTypes.func,
  systemId: PropTypes.string.isRequired
};

export default PoolColumn;
