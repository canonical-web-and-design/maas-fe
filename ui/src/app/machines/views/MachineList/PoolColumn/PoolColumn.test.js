import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import React from "react";

import PoolColumn from "./PoolColumn";

const mockStore = configureStore();

describe("PoolColumn", () => {
  let state;
  beforeEach(() => {
    state = {
      config: {
        items: []
      },
      machine: {
        errors: {},
        loading: false,
        loaded: true,
        items: [
          {
            system_id: "abc123",
            pool: { name: "default" },
            description: "Firmware old"
          }
        ]
      }
    };
  });

  it("renders", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <PoolColumn systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find("PoolColumn")).toMatchSnapshot();
  });

  it("displays pool", () => {
    state.machine.items[0].pool = { name: "pool-1" };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <PoolColumn systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('[data-test="pool"]').text()).toEqual("pool-1");
  });

  it("displays description", () => {
    state.machine.items[0].description = "decomissioned";
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <PoolColumn systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('[data-test="note"]').text()).toEqual("decomissioned");
  });
});
