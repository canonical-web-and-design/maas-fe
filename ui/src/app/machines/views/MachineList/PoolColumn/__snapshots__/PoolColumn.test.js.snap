// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoolColumn renders 1`] = `
<PoolColumn
  systemId="abc123"
>
  <DoubleRow
    menuLinks={
      Array [
        Object {
          "children": "Backup",
          "onClick": [Function],
        },
      ]
    }
    menuTitle="Change pool:"
    primary={
      <span
        data-test="pool"
      >
        <a
          className="p-link--soft"
          href="#/pools"
          title="default"
        >
          default
        </a>
      </span>
    }
    primaryAriaLabel="Pool"
    secondary={
      <span
        data-test="note"
        title="Firmware old"
      >
        Firmware old
      </span>
    }
    secondaryAriaLabel="Note"
  >
    <div
      className="p-double-row"
    >
      <div
        className="p-double-row__rows-container"
      >
        <div
          aria-label="Pool"
          className="p-double-row__primary-row"
        >
          <div
            className="p-double-row__primary-row-text u-truncate"
          >
            <span
              data-test="pool"
            >
              <a
                className="p-link--soft"
                href="#/pools"
                title="default"
              >
                default
              </a>
            </span>
          </div>
          <TableMenu
            links={
              Array [
                Object {
                  "children": "Backup",
                  "onClick": [Function],
                },
              ]
            }
            title="Change pool:"
          >
            <ContextualMenu
              className="p-table-menu"
              hasToggleIcon={true}
              links={
                Array [
                  "Change pool:",
                  Array [
                    Object {
                      "children": "Backup",
                      "onClick": [Function],
                    },
                  ],
                ]
              }
              position="center"
              toggleAppearance="base"
              toggleClassName="u-no-margin--bottom p-table-menu__toggle"
            >
              <span
                className="p-table-menu p-contextual-menu--center"
              >
                <Button
                  appearance="base"
                  aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                  aria-expanded="false"
                  aria-haspopup="true"
                  className="p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                  hasIcon={true}
                  onClick={[Function]}
                >
                  <button
                    aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                    aria-expanded="false"
                    aria-haspopup="true"
                    className="p-button--base has-icon p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                    onClick={[Function]}
                  >
                    <i
                      className="p-icon--contextual-menu p-contextual-menu__indicator"
                    />
                  </button>
                </Button>
              </span>
            </ContextualMenu>
          </TableMenu>
        </div>
        <div
          aria-label="Note"
          className="p-double-row__secondary-row u-truncate"
        >
          <span
            data-test="note"
            title="Firmware old"
          >
            Firmware old
          </span>
        </div>
      </div>
    </div>
  </DoubleRow>
</PoolColumn>
`;
